language/bytecode-verifier/src/dependencies.rs

```diff
@@ -317,6 +321,21 @@ fn verify_imported_functions(context: &Context) -> PartialVMResult<()> {
     Ok(())
 }
 
+fn compatible_type_parameters_phantom_decl(
+    local_type_parameters_declaration: &[StructTypeParameter],
+    defined_type_parameters: &[StructTypeParameter],
+) -> bool {
+    local_type_parameters_declaration.len() == defined_type_parameters.len()
+        && local_type_parameters_declaration
+            .iter()
+            .zip(defined_type_parameters)
+            .all(
+                |(local_type_parameter_declaration, defined_type_parameter)| {
+                    defined_type_parameter.is_phantom == local_type_parameter_declaration.is_phantom
```

Comment:
I think we can be looser than this? maybe?

The goal of these "compatible" checks is to make upgrades a bit more flexible without breaking anything.  Hence  the `is_subset` check in `compatible_type_parameter_constraints`

So my guess here is it is fine to drop `is_phantom` here. Something like
```
!local_type_parameter_declaration.is_phantom || defined_type_parameter.is_phantom
``` 
And you ever wish languages had logical implication? :p This would read so much better as 
```
local_type_parameter_declaration.is_phantom implies defined_type_parameter.is_phantom
```

Reply:
But I ahven't convinced myself this is safe yet

Reply:
It is safe in that a struct definition will continue to be well-formed after dropping a phantom, but it could break users of the struct. 

```
struct S<phantom T> has copy {}

struct NonCopy;
```

When dropping the phantom `S<NonCopy>` has no longer copy.

So it depends on what's the notion of compatibility here...

Reply:
On the contrary, adding a phantom could make the struct malformed but it would never break user code.

Reply:
I'm not sure what you mean by malformed. But this is precisely what I am discussing. 

In dependencies: 
- It is safe for the local definition to drop the `phantom` in it's view of the type

In upgrade/compatibility:
- It is safe for the new definition to add `phantom` in it's type (assuming all of the phantom rules apply, which are checked elsewhere)

Reply:
By well-formedness I mean phantom parameters are used in phantom position, so I think we are saying the same.